### DISCUSSION TITLE
useRouter unable to fetch query params with special characters

### DISCUSSION THREAD

**Discussion by Arpitvicky** [vercel/next.js#36853] [Help] [answered]
I faced an issue currently. I was using useRouter to get the query params

Url example:
http://localhost:3000/anything?email=anything+3003@gmail.com

Code used to fetch
```
const { query } = useRouter();
const emailId = query.email;
console.log(emailId);
```

output is -  anything 3003@gmail.com. ----> without '+'

output should be - anything+3003@gmail.com ---> with '+'

Could you please suggest a way through next js useRouter ? I can do it through JS "location.search" but it would be nice if we get something from useRouter.

Thanks in advance!

**Comment by icyJoseph** [comment, marked as answer]
You can use `asPath`, I presume there's some encoding/decoding going on when building the query object.

I tried with:

```js
  const { query, asPath } = useRouter();
  const emailId = query.email;
  console.log(emailId, asPath);

  const queryString = asPath.split("?")[1];

  const queryObj = new URLSearchParams(queryString);

  console.log(queryObj.get("email"));
```

And got the same result, missing `+`, so I am going to think that the browser API is doing something I did not know about... It must be that plus is a reserved character. Yes it is, according to: https://stackoverflow.com/q/2366260.

You could still do it manually with:

```ts
  const { asPath } = useRouter();

  const [, queryString = ""] = asPath.split("?"); 

  const queryObj = queryString
    .split("&")
    .reduce<Map<string, string>>((prev, curr) => {
      const [key, value] = curr.split("=");
      prev.set(key, value);
      return prev;
    }, new Map());

  console.log(queryObj.get("email")); // anything+3003@gmail.com
```

> SPACE is encoded as '+' or '%20'

It might be better, and easier, to trim the query value from Next router, and then replace the space with a +.

```ts
const { query } = useRouter();
const emailId = query.email || "";
console.log(emailId.trim().replace(" ", "+"));
```

**Reply by Arpitvicky** [reply]
Thanks @icyJoseph ,  the first solution worked. Regarding the second snippet, there can be many other special characters too in the query params and it can become a bit confusing to trim and replace.

**Reply by icyJoseph** [reply]
Nice.

I'd recommend reading a bit more about reserved characters, not all of them transform to whitespaces.
